library/std/src/io/copy.rs

```diff
@@ -56,12 +56,47 @@ where
 
 /// The general read-write-loop implementation of
 /// `io::copy` that is used when specializations are not available or not applicable.
-pub(crate) fn generic_copy<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> io::Result<u64>
+pub(crate) fn generic_copy<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> Result<u64>
 where
     R: Read,
     W: Write,
 {
-    let mut buf = MaybeUninit::<[u8; super::DEFAULT_BUF_SIZE]>::uninit();
+    let writer_cap = WriteBuffer::capacity(writer);
+    if writer_cap > 0 {
+        // A non-zero capacity indicates that we have a BufWriter, even though we can't assert that through types
+        // We must flush the buffer because specialization in buffered_copy will bypass BufWriter, no matter which
+        // buffer we later pass as reference
+        WriteBuffer::prepare_buffer(writer)?;
+
+        if writer_cap >= super::DEFAULT_BUF_SIZE {
```

Comment:
It seems strange to flush anyway even if writer_cap is less than DEFUALT_BUF_SIZE.

Reply:
That is explained in the comment above though. I agree it's not ideal but with the current approach it would lead to the buffered data getting lost